File: src/org/brooks/aoc/y2023/Day05.java
package org.brooks.aoc.y2023;

import org.brooks.aoc.FileUtil;
import org.brooks.aoc.Solution;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.LongStream;
import java.util.stream.Stream;

public class Day05 {
    private record Transform(
            long sourceStart,
            long sourceEnd,
            long transform
    ) {
        public static Transform create(long destStart, long sourceStart, long range) {
            return new Transform(
                    sourceStart,
                    sourceStart + range,
                    destStart - sourceStart
            );
        }

        public boolean canTransform(long num) {
            return num >= sourceStart && num < sourceEnd;
        }

        public long transform(long num) {
            return num + transform;
        }
    }

    private record Terraform(
            String name,
            List<Transform> transforms
    ) {
    }

    private record SeedInfo(
            long starting,
            long count
    ) {
        private Stream<Long> getSeeds() {
            return LongStream.range(starting, starting + count).boxed();
        }
    }

    private static List<Terraform> getAllTransformations(List<String> contents) {
        final List<Terraform> terraforms = new ArrayList<>();
        Terraform t = new Terraform("temp", new ArrayList<>());
        for (int i = 2; i < contents.size(); i++) {
            if (contents.get(i).endsWith(":")) {
                //beginning of Terraform
                t = new Terraform(contents.get(i), new ArrayList<>());
            } else if (contents.get(i).isBlank()) {
                //end of terraform
                terraforms.add(t);
            } else {
                List<Long> numbers = Stream.of(contents.get(i).split(" +")).map(Long::parseLong).toList();

                t.transforms.add(Transform.create(numbers.get(0), numbers.get(1), numbers.get(2)));
            }
        }
        terraforms.add(t);
        return terraforms;
    }

    public static class Phase1 implements Solution<Long> {
        @Override
        public Long solve(Stream<String> input) {
            List<String> contents = input.toList();

            List<Long> seeds = Stream.of(contents.get(0)
                            .substring(7)
                            .split(" +"))
                    .map(Long::parseLong)
                    .toList();

            final List<Terraform> terraforms = getAllTransformations(contents);

            return seeds.stream().mapToLong(s -> {
                long num = s;
                for (Terraform t : terraforms) {
                    final long current = num;
                    List<Transform> allMatch = t.transforms.stream().filter(o -> o.canTransform(current)).toList();
                    if (allMatch.isEmpty()) {
                        continue;
                    }
                    if (allMatch.size() > 1) {
                        System.out.println("MULTIPLE MATCH!!");
                        continue;
                    }
                    num = allMatch.get(0).transform(num);
                }
                return num;
            }).min().orElse(0L);
        }
    }

    public static class Phase2 implements Solution<Long> {
        @Override
        public Long solve(Stream<String> input) {
            List<String> contents = input.toList();

            List<Long> seeds = Stream.of(contents.get(0)
                            .substring(7)
                            .split(" +"))
                    .map(Long::parseLong)
                    .toList();
            List<SeedInfo> seedData = new ArrayList<>();
            for (int i = 0; i < seeds.size(); ) {
                seedData.add(new SeedInfo(seeds.get(i), seeds.get(i + 1)));
                i += 2;
            }

            final List<Terraform> terraforms = getAllTransformations(contents);

            final AtomicLong a = new AtomicLong(0L);

            return seedData.parallelStream().flatMap(SeedInfo::getSeeds).mapToLong(s -> {
                long num = s;
                for (Terraform t : terraforms) {
                    final long current = num;
                    List<Transform> allMatch = t.transforms.stream().filter(o -> o.canTransform(current)).toList();
                    if (allMatch.isEmpty()) {
                        continue;
                    }
                    if (allMatch.size() > 1) {
                        System.out.println("MULTIPLE MATCH!!");
                        continue;
                    }
                    num = allMatch.get(0).transform(num);
                }
                long at = a.addAndGet(1L);
                if (at%10000000 == 0){
                    System.out.printf("Processed %,d records, %s\n", at, Thread.currentThread().getName());
                }
                return num;
            }).reduce(Long.MAX_VALUE, Math::min);
        }
    }

    public static void main(String[] args) {
        FileUtil.runInputThroughFunction("Phase 1 - Example -", "2023/05-Example-01.txt", new Phase1());
        FileUtil.runInputThroughFunction("Phase 1 - Input -", "2023/05.txt", new Phase1());
        FileUtil.runInputThroughFunction("Phase 2 - Example -", "2023/05-Example-01.txt", new Phase2());
        FileUtil.runInputThroughFunction("Phase 2 - Input -", "2023/05.txt", new Phase2());
    }
}

File: y2023/day05.go
package main

import (
	"bufio"
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/lbrooks/aoc/utils"
)

type Transform struct {
	sourceStart, sourceEnd, change int64
}

func (t Transform) canTransform(num int64) bool {
	return num >= t.sourceStart && num < t.sourceEnd
}

func (t Transform) transform(num int64) int64 {
	return num + t.change
}

func NewTransform(destStart, sourceStart, rangeLength int64) Transform {
	return Transform{
		sourceStart: sourceStart,
		sourceEnd:   sourceStart + rangeLength,
		change:      destStart - sourceStart,
	}
}

type Terraform struct {
	name       string
	transforms []Transform
}

type SeedInfo struct {
	starting, count int64
}

func (s SeedInfo) getSeeds() []int64 {
	result := make([]int64, s.count)
	for i := 0; i < int(s.count); i++ {
		result[i] = s.starting + int64(i)
	}
	return result
}

func getAllTransformations(contents []string) ([]int64, []Terraform) {
	seeds := make([]int64, 0)
	terras := make([]Terraform, 0)

	re := regexp.MustCompile((" +"))
	sas := re.Split(contents[0][7:], -1)
	for _, s := range sas {
		n, _ := strconv.ParseInt(s, 10, 64)
		seeds = append(seeds, n)
	}

	t := Terraform{"", nil}
	for i := 2; i < len(contents); i++ {
		if strings.HasSuffix(contents[i], ":") {
			t = Terraform{contents[i], nil}
		} else if len(contents[i]) == 0 {
			terras = append(terras, t)
		} else {
			numbs := strings.Split(contents[i], " ")

			dest, _ := strconv.ParseInt(strings.Trim(numbs[0], " "), 10, 64)
			start, _ := strconv.ParseInt(strings.Trim(numbs[1], " "), 10, 64)
			count, _ := strconv.ParseInt(strings.Trim(numbs[2], " "), 10, 64)

			t.transforms = append(t.transforms, NewTransform(dest, start, count))
		}
	}
	terras = append(terras, t)
	return seeds, terras
}

func readInputToSlice(input *bufio.Scanner) (contents []string) {
	for input.Scan() {
		contents = append(contents, input.Text())
	}
	return
}

type Day05_Phase1 struct{}

func (d Day05_Phase1) Solve(input *bufio.Scanner) (int64, error) {
	contents := readInputToSlice(input)

	seeds, terraforms := getAllTransformations(contents)

	min, minSet := int64(0), false
	for _, s := range seeds {
		workingNum := s
		for _, t := range terraforms {
			for _, tf := range t.transforms {
				if tf.canTransform(workingNum) {
					workingNum = tf.transform(workingNum)
					break
				}
			}
		}
		if !minSet {
			min = workingNum
			minSet = true
		} else if workingNum < min {
			min = workingNum
		}
	}
	return min, nil
}

type Day05_Phase2 struct{}

func (d Day05_Phase2) Solve(input *bufio.Scanner) (int64, error) {
	contents := readInputToSlice(input)

	rawSeedData, terraforms := getAllTransformations(contents)

	sis := make([]SeedInfo, 0)
	for i := 0; i < len(rawSeedData); i = i + 2 {
		sis = append(sis, SeedInfo{rawSeedData[0], rawSeedData[1]})
	}

	min, minSet := int64(0), false
	for id, si := range sis {
		for i := int64(0); i < si.count; i++ {
			workingNum := si.starting + i
			for _, t := range terraforms {
				for _, tf := range t.transforms {
					if tf.canTransform(workingNum) {
						workingNum = tf.transform(workingNum)
						break
					}
				}
			}
			if !minSet {
				min = workingNum
				minSet = true
			} else if workingNum < min {
				min = workingNum
			}
		}
		fmt.Printf("Completed %d of %d seeds\n", id+1, len(sis))
	}
	return min, nil
}

func main() {
	err := utils.RunInputThroughFunction("Phase 1 Example", "../resources/2023/05-Example-01.txt", &Day05_Phase1{})
	if err != nil {
		fmt.Println("Error: %s", err.Error())
	}

	err = utils.RunInputThroughFunction("Phase 1", "../resources/2023/05.txt", &Day05_Phase1{})
	if err != nil {
		fmt.Println("Error: %s", err.Error())
	}

	err = utils.RunInputThroughFunction("Phase 2 Example", "../resources/2023/05-Example-01.txt", &Day05_Phase2{})
	if err != nil {
		fmt.Println("Error: %s", err.Error())
	}

	err = utils.RunInputThroughFunction("Phase 2", "../resources/2023/05.txt", &Day05_Phase2{})
	if err != nil {
		fmt.Println("Error: %s", err.Error())
	}
}
